pkg/server/handler.go

```diff
@@ -525,44 +539,68 @@ func (al *activeListener) activeStreamSize() int {
 	return int(s.Counter(metrics.DownstreamRequestActive).Count())
 }
 
+var (
+	// drain time, default 15 seconds
+	drainTime = time.Second * 15
+)
+
+func SetDrainTime(time time.Duration) {
+	drainTime = time
+}
+
+// OnShutdown graceful stop the existing connection and wait all connections to be closed
+func (al *activeListener) OnShutdown() {
+	utils.GoWithRecover(func() {
+		al.conns.VisitSafe(func(v interface{}) {
+			conn := v.(*activeConnection).conn
+			// TODO 1: conn.OnConnectionEvent may be blocked on connection.Write, need a proper way to not block too long.
+			// TODO 2: shutdown connections gradual, it's useful when there are many connections.
```

Comment:
yeah, it's worth resolving it.
I think it's better to put it in another PR after this PR is merged, since the PR is big enough now.
And this PR has already reached a good enough milestone.